src/processor.cpp

```diff
@@ -0,0 +1,394 @@
+
+}
+
+void processor_data::load( JsonObject &jo, const std::string & )
+{
+    mandatory( jo, was_loaded, "next_type", next_type );
+    mandatory( jo, was_loaded, "fuel_type", fuel_type );
+    mandatory( jo, was_loaded, "processes", processes );
+
+    if( !was_loaded || jo.has_member( "messages" ) ) {
+        JsonObject messages_obj = jo.get_object( "messages" );
+
+        optional( messages_obj, was_loaded, "empty", empty_message, translated_string_reader );
+        optional( messages_obj, was_loaded, "full", full_message, translated_string_reader );
+        optional( messages_obj, was_loaded, "start", start_message, translated_string_reader );
+        optional( messages_obj, was_loaded, "processing", processing_message, translated_string_reader );
+        optional( messages_obj, was_loaded, "finish", finish_message, translated_string_reader );
+        optional( messages_obj, was_loaded, "fail", fail_message, translated_string_reader );
+    }
+}
+
+void processor_data::check() const
+{
+    static const iexamine_function converter( iexamine_function_from_string( "converter" ) );
+    const furn_str_id processor_id( id.str() );
+
+    if( !processor_id.is_valid() ) {
+        debugmsg( "Furniture \"%s\" have no processor of the same name.", id.c_str() );
+    } else if( processor_id->examine != converter ) {
+        debugmsg( "Furniture \"%s\" can't act as a processor because the assigned iexamine function is not converter, but it has \"%s\" associated.",
+                  processor_id.c_str(), id.c_str() );
+    }
+    for( const std::string &elem : processes ) {
+        if( !get_process_id(elem).is_valid() ) {
+            debugmsg( "Invalid process \"%s\" in \"%s\".", elem.c_str(), id.c_str() );
+        }
+    }
+}
+
+void processors::load( JsonObject &jo, const std::string &src )
+{
+    processors_data.load( jo, src );
+}
+
+void processors::check()
+{
+    processors_data.check();
+}
+
+void processors::reset()
+{
+    processors_data.reset();
+}
+
+template <typename T>
+bool IsSubset(std::vector<T> A, std::vector<T> B)
+{
+    std::sort(A.begin(), A.end());
+    std::sort(B.begin(), B.end());
+    return std::includes(A.begin(), A.end(), B.begin(), B.end());
+}
+
+const process_data& select_active(const tripoint &examp )
+{
+    processor_id pid = get_processor_id(examp);
+    const processor_data &current_processor = processors_data.obj(pid);
+    //Make an object list of processes
+    std::vector<process_data> *possible_processes = new std::vector<process_data>();
+    for (const std::string &p : current_processor.processes) {
+        possible_processes->push_back( processes_data.obj( get_process_id(p) ));
+    }
+    //Figure out active process
+    process_data *active_process = nullptr;
+    if (possible_processes->size() == 1) {
+        *active_process = possible_processes->at(0);
+    }
+    else {
+        map_stack items = g->m.i_at( examp );
+        if (!items.empty()) {
```

Comment:
It would be cleaner if you inverted this check.
As in,
```
if( items.empty() ) {
    return nullptr;
}
```

Then you wouldn't have this big `if/else` ladder.